rlkit/core/rl_algorithm.py

```diff
@@ -238,74 +220,31 @@ def pretrain(self):
-
-    def collect_data_from_task_posterior(self, idx, num_samples=1, resample_z_every_n=None, eval_task=False,
-                                         add_to_enc_buffer=True):
-        # do not resample z if resample_z_every_n is None
-        if resample_z_every_n is None:
-            self.sample_z_from_posterior(idx, eval_task=eval_task)
-            self.collect_data(self.policy, num_samples=num_samples, eval_task=eval_task,
-                              add_to_enc_buffer=add_to_enc_buffer)
-        else:
-            # collects more data in batches of resample_z_every_n until done
-            while num_samples > 0:
-                self.collect_data_from_task_posterior(idx=idx,
-                                                      num_samples=min(resample_z_every_n, num_samples),
-                                                      resample_z_every_n=None,
-                                                      eval_task=eval_task,
-                                                      add_to_enc_buffer=add_to_enc_buffer)
-                num_samples -= resample_z_every_n
-
-    # split number of prior and posterior samples
-    def collect_data_online(self, idx, num_samples, eval_task=False, add_to_enc_buffer=True):
-        self.collect_data_sampling_from_prior(num_samples=num_samples,
-                                              resample_z_every_n=self.max_path_length,
-                                              eval_task=eval_task,
-                                              add_to_enc_buffer=True)
-        self.collect_data_from_task_posterior(idx=idx,
-                                              num_samples=num_samples,
-                                              resample_z_every_n=self.max_path_length,
-                                              eval_task=eval_task,
-                                              add_to_enc_buffer=add_to_enc_buffer)
-
+    def collect_data(self, num_samples, resample_z_rate, update_posterior_rate, update_batch_size, add_to_enc_buffer=True):
+        '''
+        collect transitions from the task
+        :param task_idx: the task ID to sample from
+        :param num_samples: total number of transitions to sample
+        :param resample_z_rate: how often to resample latent context z
+        :param update_posterior_rate: how often to update q(z | c) from which z is sampled
+        :param update_batch_size: controls what data is used to infer the new q(z | c)
+        '''
+        # makes no sense to update the posterior without resampling z
+        assert resample_z_rate <= update_posterior_rate
+        # start from the prior
+        self.reset_posterior()
+        counter = 0
+        for _ in range(num_samples // resample_z_rate):
+            self.collect_transitions(self.policy, num_samples=resample_z_rate, add_to_enc_buffer=add_to_enc_buffer)
+            counter += resample_z_rate
+            if counter % update_posterior_rate == 0:
```

Comment:
If for some reason resample_z_rate does not divide update_posterior_rate, this doesn't work as it is probably intended. Also, given that we reset the envs between switching tasks, maybe collect transitions should just collect complete trajectories until we exceed num_samples (returning the # of smaples it consumed), and then update posterior if stepcount >= update_count * posterior rate.

Reply:
see `81f27a8`